cpp/include/cudf/detail/aggregation/aggregation.hpp

```diff
@@ -0,0 +1,324 @@
+  std::vector<double> _quantiles;              ///< Desired quantile(s)
+  experimental::interpolation _interpolation;  ///< Desired interpolation
+};
+
+/**
+ * @brief Sentinel value used for `ARGMAX` aggregation.
+ *
+ * The output column for an `ARGMAX` aggregation is initialized with the
+ * sentinel value to indicate an unused element.
+ */
+constexpr size_type ARGMAX_SENTINEL{-1};
+
+/**
+ * @brief Sentinel value used for `ARGMIN` aggregation.
+ *
+ * The output column for an `ARGMIN` aggregation is initialized with the
+ * sentinel value to indicate an unused element.
+ */
+constexpr size_type ARGMIN_SENTINEL{-1};
+
+/**	
+ * @brief Maps an `aggregation::Kind` value to it's corresponding binary	
+ * operator.	
+ *	
+ * @note Not all values of `aggregation::Kind` have a valid corresponding binary	
+ * operator. For these values `E`,	
+ * `std::is_same_v<corresponding_operator<E>::type, void>`.	
+ *	
+ * @tparam k The `aggregation::Kind` value to map to its corresponding operator	
+ */	
+template <aggregation::Kind k>	
+struct corresponding_operator {	
+    using type = void;	
+};	
+
+template <>	
+struct corresponding_operator<aggregation::MIN> {	
+  using type = DeviceMin;	
+};	
+template <>	
+struct corresponding_operator<aggregation::MAX> {	
+  using type = DeviceMax;	
+};	
+template <>	
+struct corresponding_operator<aggregation::SUM> {	
+  using type = DeviceSum;	
+};	
+
+template <aggregation::Kind k>	
+using corresponding_operator_t = typename corresponding_operator<k>::type;	
+
+/**---------------------------------------------------------------------------*
+ * @brief Determines accumulator type based on input type and aggregation.
+ *
+ * @tparam Source The type on which the aggregation is computed
+ * @tparam k The aggregation performed
+ *---------------------------------------------------------------------------**/
+template <typename Source, aggregation::Kind k, typename Enable = void>
+struct target_type_impl {
+  using type = void;
+};
+
+// Computing MIN of Source, use Source accumulator
+template <typename Source>
+struct target_type_impl<Source, aggregation::MIN> {
+  using type = Source;
+};
+
+// Computing MAX of Source, use Source accumulator
+template <typename Source>
+struct target_type_impl<Source, aggregation::MAX> {
+  using type = Source;
+};
+
+// Always use size_type accumulator for COUNT
+template <typename Source>
+struct target_type_impl<Source, aggregation::COUNT> {
+  using type = cudf::size_type;
+};
+
+// Always use `double` for MEAN
+template <typename Source>
+struct target_type_impl<Source, aggregation::MEAN> {
+  using type = double;
+};
+
+// Summing integers of any type, always use int64_t accumulator
+template <typename Source>
+struct target_type_impl<Source, aggregation::SUM,
+                        std::enable_if_t<std::is_integral<Source>::value>> {
+  using type = int64_t;
+};
+
+// Summing float/doubles, use same type accumulator
```

Comment:
Shouldn't this be double as there is possibility that they can cross float boundary?

Reply:
Summing with `float` instead of `double` will just result in less precision. It was decided if the user wants to sum in `double` they can cast the column ahead of time.